README.md
node-spotify-web
================
### Node.js implementation of the Spotify Web protocol

This module implements the "Spotify Web" WebSocket protocol that is used on
Spotify's [Web UI](http://play.spotify.com).

This module is heavily inspired by the original open-source Python implementation:
[Hexxeh/spotify-websocket-api](https://github.com/Hexxeh/spotify-websocket-api).

Installation
------------

``` bash
$ npm install spotify-web
```


Example
-------

Here's an example of logging in to the Spotify server and creating a session. Then
requesting the metadata for a given track URI, and playing the track audio file
through the speakers:

``` javascript
var lame = require('lame');
var Speaker = require('speaker');
var Spotify = require('spotify-web');
var uri = process.argv[2] || 'spotify:track:6tdp8sdXrXlPV6AZZN2PE8';

// Spotify credentials...
var username = process.env.USERNAME;
var password = process.env.PASSWORD;

Spotify.login(username, password, function (err, spotify) {
  if (err) throw err;

  // first get a "Track" instance from the track URI
  spotify.get(uri, function (err, track) {
    if (err) throw err;
    console.log('Playing: %s - %s', track.artist[0].name, track.name);

    // play() returns a readable stream of MP3 audio data
    track.play()
      .pipe(new lame.Decoder())
      .pipe(new Speaker())
      .on('finish', function () {
        spotify.disconnect();
      });

  });
});
```

See the `example` directory for some more example code.


API
---

TODO: document!


History.md

1.3.0 / 2014-12-11
==================

  * spotify: update comment to explain the web service
  * spotify: use `ping-pong.spotify.nodestuff.net` web service for sendPong (#98, @denysvitali)
  * spotify: set "User-Agent" and "Origin" on WebSocket connection (#96, @denysvitali)
  * add LICENSE file
  * README: add License section
  * s/NodeJS/Node.js/
  * example: add an example of getting Artist "genre"

1.2.0 / 2014-04-26
==================

  * spotify: new flash key (#81, @brandtabbott)
  * spotify: added functionality to get a users starred playlist with example

1.1.1 / 2014-04-02
==================

  * example: switch
  * spotify: added `Spotify.Web.App.initialize()` noop Function (#76)

1.1.0 / 2014-02-24
==================

  * spotify: handle `sp/ping_flash2` commands
  * spotify: emit "open" and "close" events
  * spotify: fix lint
  * spotify: use client version from auth response

1.0.0 / 2014-01-08
==================

  * spotify: update user agent and send window size log event on connection (#60)
  * spotify: tag function bugfix
  * proto: update schemas module api, add protobufjs support
  * proto: add pubsub and playlist4service
  * proto: Update metadata protobuf with added Catalogue SHUFFLE enum
  * bugfix for multiGet
  * spotify: use http status message in error when none is defined
  * spotify: add similar(uri, fn) method
  * proto: add bartender schemas
  * track: cache previewUrl from getter in playPreview method and emit error on stream in nextTick
  * track: emit error if preview stream does not respond with 200 status code
  * spotify: fix bug which prevented facebook and anonymous login from working
  * spotify: add facebook and anonymous login
  * spotify: improve authentication and include trackingId
  * spotify: add XXX comment...
  * spotify: more robust `has()` function
  * track: emit error if stream does not respond with 200 status code
  * spotify: use correct number argument for sending SUB/UNSUB MercuryRequests
  * spotify: make callback of sendProtobufRequest optional
  * track: add previewUrl getter and playPreview method
  * example: add preview playing example
  * proto: update metadata fields
  * spotify: use SelectedListContent instead of ListDump for playlist and rootlist responseSchemas
  * spotify: refactor MercuryRequest code into sendProtobufRequest function

0.1.3 / 2013-06-16
==================

  * proto: change contentType to bytes format
  * spotify: support MercuryMultiGetRequest in `get()` function
  * util: fix comment
  * spotify: fix lint

0.1.2 / 2013-05-18
==================

  * spotify: use AP resolver to connect to websocket server (GH-13) @adammw

0.1.1 / 2013-03-22
==================

  * error: add error code for non-premium accounts

0.1.0 / 2013-03-09
==================

  * spotify: implement real error handling
  * spotify: ignore "login_complete" message commands
  * spotify: throw an error on unhandled "message" commands
  * error: add SpotifyError class
  * spotify: make rootlist() user default to yourself
  * track: send the User-Agent in .play()
  * Added `rootlist()` function to get a user's stored playlists

0.0.2 / 2013-02-07
==================

  * Fix CSRF token retrieval
  * A whole lot of API changes... too much to list...

0.0.1 / 2013-01-12
==================

  * Initial release:
    * getting Artist/Album/Track metadata works
    * getting MP3 playback URL for a Track works


lib/uri.js

/**
 * Module dependencies.
 */

var base62 = require('./base62');
var debug = require('debug')('spotify-web:uri');

/**
 * Module exports.
 */

module.exports = SpotifyUri;

SpotifyUri.PLAYLIST_TYPES = ['playlist', 'starred', 'rootlist', 'publishedrootlist'];

/**
 * Create a new SpotifyUri instance from a given uri type and gid
 * 
 * @param {String} uriType
 * @param {Buffer} gid
 */

SpotifyUri.fromGid = function(uriType, gid) {
  return new SpotifyUri(SpotifyUri.gid2uri(uriType, gid));
};

/**
 * Create a new SpotifyUri instance from a given uri type and id
 * 
 * @param {String} uriType
 * @param {String} id (hexadecimal)
 */

SpotifyUri.fromId = function(uriType, id) {
  return new SpotifyUri(SpotifyUri.id2uri(uriType, gid));
};

/**
 * Create a new SpotifyUri instance from a given uri
 * 
 * @param {String} uri
 */

SpotifyUri.fromUri = function(uri) {
  return new SpotifyUri(uri);
};

/**
 * SpotifyUri class.
 *
 * @api public
 */

function SpotifyUri(type, id) {
  if ('string' != typeof type) throw new Error('Invalid URI type');

  this._uri_parts = [];

  // TODO(adammw): support playlists in constructor

  if (!id) {
    this.uri = type;
  } else {
    if (id instanceof Buffer) id = SpotifyUri.gid2id(id);
    if (/^[0-9a-f]*$/.test(id)) id = base62.fromHex(id, 22);
    this._uri_parts = ['spotify', type, id];
  }
}

/**
 * SpotifyUri uri getter / setter
 */

Object.defineProperty(SpotifyUri.prototype, 'uri', {
  get: function () {
    var uri = this._uri_parts.join(':');
    debug('get uri() : %s', uri)
    return uri;
  },
  set: function (uri) {
    debug('set uri() : %s', uri);
    var uri_parts = uri.split(':');
    if ('spotify' != uri_parts[0]) throw new Error('Invalid Spotify Uri');
    this._uri_parts = uri_parts;
  },
  enumerable: true,
  configurable: true
});

/**
 * SpotifyUri type getter
 */

Object.defineProperty(SpotifyUri.prototype, 'type', {
  get: function () {
    var parts = this._uri_parts;
    var len = parts.length;

    if (len >= 3 && 'local' == parts[1]) {
      // e.g. spotify:local:AC%2FDC:Highway+to+Hell:Highway+to+Hell:209
      return 'local';
    } else if (len >= 5) {
      // e.g. spotify:user:tootallnate:[playlist]:0Lt5S4hGarhtZmtz7BNTeX
      return parts[3];
    } else if (len >= 4 && SpotifyUri.PLAYLIST_TYPES.indexOf(parts[3]) !== -1) {
      // e.g. spotify:user:tootallnate:starred
      return parts[3];
    } else if (len >= 3) {
      // e.g. spotify:[track]:6tdp8sdXrXlPV6AZZN2PE8
      return parts[1];
    } else {
      return null;
    }
  },
  enumerable: true,
  configurable: true
});

/**
 * SpotifyUri sid getter / setter
 *
 * e.g. '6tdp8sdXrXlPV6AZZN2PE8' for spotify:track:6tdp8sdXrXlPV6AZZN2PE8
 */

Object.defineProperty(SpotifyUri.prototype, 'sid', {
  get: function () {
    var parts = this._uri_parts;
    var len = parts.length;

    return parts[len - 1];
  },
  set: function (sid) {
    var parts = this._uri_parts;
    var len = parts.length;

    parts[len - 1] = sid;
  },
  enumerable: true,
  configurable: true
});

/**
 * SpotifyUri id getter / setter
 *
 * e.g. 'd49fcea60d1f450691669b67af3bda24' for spotify:track:6tdp8sdXrXlPV6AZZN2PE8
 */

Object.defineProperty(SpotifyUri.prototype, 'id', {
  get: function () {
    return base62.toHex(this.sid);
  },
  set: function (id) {
    this.sid = base62.fromHex(id, 22);
  },
  enumerable: true,
  configurable: true
});

/**
 * SpotifyUri gid getter / setter
 *
 * e.g. <Buffer d4 9f ce a6 0d 1f 45 06 91 66 9b 67 af 3b da 24> for spotify:track:6tdp8sdXrXlPV6AZZN2PE8
 */

Object.defineProperty(SpotifyUri.prototype, 'gid', {
  get: function () {
    return new Buffer(this.id, 'hex');
  },
  set: function (gid) {
    this.id = gid.toString('hex');
  },
  enumerable: true,
  configurable: true
});

/**
 * SpotifyUri user getter / setter
 */

Object.defineProperty(SpotifyUri.prototype, 'user', {
  get: function () {
    var parts = this._uri_parts;
    var len = parts.length;

    if (len >= 3 && 'user' == parts[1]) return parts[2];
    return null;
  },
  set: function (user) {
    var parts = this._uri_parts;
    var len = parts.length;

    if (len >= 3 && 'user' == parts[1]) parts[2] = user;
  },
  enumerable: true,
  configurable: true
});

/**
 * Returns the underlying uri string
 *
 * @return {String}
 */
SpotifyUri.prototype.toString = function() {
  return this.uri;
};

/**
 * Converts a GID Buffer to an ID hex string.
 * Provided for backwards compatibility.
 */

SpotifyUri.gid2id = function (gid) {
  return gid.toString('hex');
};

/**
 * ID -> URI
 * Provided for backwards compatibility.
 */

SpotifyUri.id2uri = function (uriType, id) {
  return (new SpotifyUri(uriType, id)).uri;
};

/**
 * URI -> ID
 * Provided for backwards compatibility.
 *
 * >>> SpotifyUtil.uri2id('spotify:track:6tdp8sdXrXlPV6AZZN2PE8')
 * 'd49fcea60d1f450691669b67af3bda24'
 * >>> SpotifyUtil.uri2id('spotify:user:tootallnate:playlist:0Lt5S4hGarhtZmtz7BNTeX')
 * '192803a20370c0995f271891a32da6a3'
 */

SpotifyUri.uri2id = function (uri) {
  return (uri instanceof SpotifyUri) ? uri.id : (new SpotifyUri(uri)).id;
};

/**
 * GID -> URI
 * Provided for backwards compatibility.
 */

SpotifyUri.gid2uri = function (uriType, gid) {
  return (new SpotifyUri(uriType, gid)).uri;
};

/**
 * Accepts a String URI, returns the "type" of URI.
 * i.e. one of "local", "playlist", "track", etc.
 *
 * Provided for backwards compatibility.
 */

SpotifyUri.uriType = function (uri) {
  return (uri instanceof SpotifyUri) ? uri.type : (new SpotifyUri(uri)).type;
};
